### PR TITLE
added the documentation for app context.

### PR DESCRIPTION
**Pull request by amosmachora**
This PR adds an initial description of the definition and usage of the app context, as I understand it.

### CODE REVIEW

File: pages/docs/recipes/store-values.en-US.mdx [hunk 1]
```diff
@@ -86,6 +86,38 @@ userStore.set((state) => state.user = /* value from somewhere */);
 
 Updating the store using `set()` will update all variables subscribed to the store, including (like in the example) `user`.
 
+#### App context
+
+Provided by `@openmrs/esm-framework` it`s another way to store data in O3. If you are familiar with [React's context](https://react.dev/reference/react/useContext#reference) it works similarly in the same way. All you
+need is to create a global context in some higher level component and access it somewhere in any of its children.
+The `useDefineAppContext` hook takes in the name of your context and its initial values.
```
Comment (chibongho) [inline pages/docs/recipes/store-values.en-US.mdx:93]:
```suggestion
Provided by `@openmrs/esm-framework` it's another way to store data in O3. If you are familiar with [React's context](https://react.dev/reference/react/useContext#reference), it works similarly. Unlike React's Context, App Context works
across React subtrees, and thus can work in extensions and workspaces. All you need is to create a global context in
some higher level component and access it somewhere in any of its children. The `useDefineAppContext` hook takes in
the name of your context and its initial values.
```

**Review by chibongho — APPROVED** [review]
Thanks @amosmachora . Looks good. Added a suggestion on highlighting difference between our AppContext and React's Context

File: pages/docs/recipes/store-values.en-US.mdx (hunk 1 above)
Comment (amosmachora) [reply pages/docs/recipes/store-values.en-US.mdx]: Thanks @chibongho I like the suggestion 👍 .

File: pages/docs/recipes/store-values.en-US.mdx [hunk 2]
```diff
@@ -86,6 +86,39 @@ userStore.set((state) => state.user = /* value from somewhere */);
 
 Updating the store using `set()` will update all variables subscribed to the store, including (like in the example) `user`.
 
+#### App context
+
+Provided by `@openmrs/esm-framework` it's another way to store data in O3. If you are familiar with [React's context](https://react.dev/reference/react/useContext#reference), it works similarly. Unlike React's Context, App Context works
+across React subtrees, and thus can work in extensions and workspaces. All you need is to create a global context in
+some higher level component and access it somewhere in any of its children. The `useDefineAppContext` hook takes in
+the name of your context and its initial values.
```
Comment (denniskigen) [inline pages/docs/recipes/store-values.en-US.mdx:94]:
```suggestion
[App Context](https://github.com/openmrs/openmrs-esm-core/pull/976) is an application-wide context that can be used to store data that is shared across different parts of the application. To leverage it, you need to define a context with a name and initial values, and then access it somewhere in any of its children. The framework provides two hooks to make it easier:

- [useDefineAppContext](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-react-utils/src/useDefineAppContext.ts) for defining the context.
- [useAppContext](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-react-utils/src/useAppContext.ts) for accessing the context.
```

File: pages/docs/recipes/store-values.en-US.mdx [hunk 3]
```diff
@@ -86,6 +86,39 @@ userStore.set((state) => state.user = /* value from somewhere */);
...
+const [dateRange, setDateRange] = useState<Date[]>([
+  dayjs().startOf("day").toDate(),
+  new Date(),
+]);
+useDefineAppContext<DateFilterContext>("laboratory-date-filter", {
+  dateRange,
+  setDateRange,
+});
+```
+
```
Comment (denniskigen) [inline pages/docs/recipes/store-values.en-US.mdx:108]:
```suggestion
##### Defining a context

```tsx
const [dateRange, setDateRange] = useState<Date[]>([
  dayjs().startOf("day").toDate(),
  new Date(),
]);
useDefineAppContext<DateFilterContext>("laboratory-date-filter", {
  dateRange,
  setDateRange,
});
```

```

File: pages/docs/recipes/store-values.en-US.mdx [hunk 4]
```diff
@@ -86,6 +86,39 @@ userStore.set((state) => state.user = /* value from somewhere */);
...
+const { dateRange, setDateRange } = useAppContext<DateFilterContext>(
+  "laboratory-date-filter"
+) ?? {
+  dateRange: "fallback-date-range",
+  setDateRange: "fallback setter function",
+};
+```
+
+the fallback object after the `nullish coalescing` check is for providing a fallback value when the app context is undefined.
+
```
Comment (denniskigen) [inline pages/docs/recipes/store-values.en-US.mdx:121]:
```suggestion
##### Accessing the context

```tsx
const { dateRange, setDateRange } = useAppContext<DateFilterContext>(
  "laboratory-date-filter"
) ?? {
  // Fallback values used when the app context is undefined
  dateRange: "fallback-date-range",
  setDateRange: "fallback setter function",
};
```

```

**Comment by ibacher** [comment]
Hmmm... I think we need to more clearly differentiate "global stores" from the "app context". The idea of the app context is specifically to have shared state that only lives as long as a specific component remains part of the active application, basically to emulate shared context, but in a way that works across different React trees. So it's useful when you have a piece of state in one component that needs to be shared with other components, e.g., to store things like "this is the patient we're working on" vs global stores, which will be retained between component renders.

All of the above there is probably too much technical gobbledygook to be included as-is, but I feel that the docs should give users some way of understanding when you'd chose global stores vs the app context vs other possible persistence solutions.

File: pages/docs/recipes/store-values.en-US.mdx [hunk 5]
```diff
@@ -88,7 +88,14 @@ Updating the store using `set()` will update all variables subscribed to the sto
 
 #### App context
 
-[App Context](https://github.com/openmrs/openmrs-esm-core/pull/976) is an application-wide context that can be used to store data that is shared across different parts of the application. To leverage it, you need to define a context with a name and initial values, and then access it somewhere in any of its children. The framework provides two hooks to make it easier:
+[App Context](https://github.com/openmrs/openmrs-esm-core/pull/976) is a mechanism for sharing state across different parts of the application, specifically designed for cross-component communication within the currently active application. Unlike global stores which persist between component renders, App Context is tied to the lifecycle of specific components.
```
Comment (denniskigen) [inline pages/docs/recipes/store-values.en-US.mdx:91]: @ibacher do these edits bring out that distinction to some degree?

**Pull request merged**